### PR TITLE
audit(9): Round gas adjustments in favor of swapper

### PR DESCRIPTION
**Pull request by alanhwu**
Round gas adjustments in favor of swappers based on input/output and gas price increasing/decreasing

### CODE REVIEW

File: src/reactors/V3DutchOrderReactor.sol [hunk 1]
```diff
@@ -112,15 +112,19 @@ contract V3DutchOrderReactor is BaseReactor {
         // positive means an increase in gas
         int256 gasDeltaGwei = block.basefee.sub(order.startingBaseFee);
 
-        // Gas increase should increase input
-        int256 inputDelta = int256(order.baseInput.adjustmentPerGweiBaseFee) * gasDeltaGwei / 1 gwei;
+        // Gas increase should increase input (round up when positive, down when negative)
```
Comment (zhongeric) [inline src/reactors/V3DutchOrderReactor.sol:115]: nit: add comment here about how less input = favoring swapper

File: test/reactors/V3DutchOrderReactor.t.sol [hunk 2]
```diff
@@ -1401,6 +1401,54 @@ contract V3DutchOrderTest is PermitSignature, DeployPermit2, BaseReactorTest {
...
+        // Test smaller gas changes
+        vm.fee(1.1 gwei);
+        resolvedOrder = quoter.quote(order.order, order.sig);
+        assertEq(resolvedOrder.input.amount, 100 ether + 0.08 gwei, "Input should handle small positive gas changes");
+        assertEq(resolvedOrder.outputs[0].amount, 100 ether - 0.08 gwei, "Output should handle small positive gas changes");
+
+        vm.fee(0.9 gwei);
+        resolvedOrder = quoter.quote(order.order, order.sig);
+        assertEq(resolvedOrder.input.amount, 100 ether - 0.08 gwei, "Input should handle small negative gas changes");
+        assertEq(resolvedOrder.outputs[0].amount, 100 ether + 0.08 gwei, "Output should handle small negative gas changes");
```
Comment (zhongeric) [inline test/reactors/V3DutchOrderReactor.t.sol:1449]: is there a test that specifically tests the rounding down / up?

Comment (alanhwu) [reply test/reactors/V3DutchOrderReactor.t.sol]: accidentally tested using gwei instead of wei. fixed

File: src/reactors/V3DutchOrderReactor.sol [hunk 3]
```diff
@@ -109,22 +110,32 @@ contract V3DutchOrderReactor is BaseReactor {
...
         }
     }
 
+    function _computeDelta(uint256 adjustmentPerGweiBaseFee, int256 gasDeltaWei) internal pure returns (int256) {
+        if (gasDeltaWei >= 0) {
+            // Gas increase: round adjustment down to decrease amount added to input or subtracted from output
+            return int256(adjustmentPerGweiBaseFee.mulDivDown(uint256(gasDeltaWei), 1 gwei));
+        } else {
+            // Gas decrease: round adjustment up to increase amount added to output or subtracted from input
+            return -int256(adjustmentPerGweiBaseFee.mulDivUp(SafeCast.toUint256(-gasDeltaWei), 1 gwei));
```
Comment (zhongeric) [inline src/reactors/V3DutchOrderReactor.sol:135]: why do you need SafeCast here? does "unit256(-gasDeltaWei)" not work?

**Review by zhongeric — APPROVED** [review]

**Review by codyborn — APPROVED** [review]
Looks good. I'm surprised that this lowered the gas costs!

**Review by codyborn — APPROVED** [review]

**Pull request merged**